### PR TITLE
OPS-16137-Do not spam with 4xx CF errors in #alerts Slack channel

### PR DESCRIPTION
**Pull request by cristi-crunchyroll**
# Ticket
[OPS-16137](https://ellation.atlassian.net/browse/OPS-16137)

# Details
Remove the 4xx cloudfront errors from ef-generate

### CODE REVIEW

**Review by csalagean-ellation — CHANGES_REQUESTED** [review]
I will not delete the condition, but I'll increase the threshold for 4xx.

**Comment by cristi-crunchyroll** [comment]
I have filtered the distros considering only those that have nr_monitoring set to enabled.

File: efopen/newrelic_executor.py [hunk 1]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
-    map_function = lambda x: (x['Id'], ', '.join(x['Aliases']['Items']))
-    queue = map(map_function, distribution_list['Items'])
-
-    while distribution_list['IsTruncated']:
-      distribution_list = cloudfront.list_distributions(Marker=distribution_list['NextMarker'])['DistributionList']
+    paginator = cloudfront.get_paginator('list_distributions')
+    pages = paginator.paginate()
+    queue = []
+    for page in pages:
+      map_function = lambda x: (x['Id'], ', '.join(x['Aliases']['Items']))
```
Comment (csalagean-ellation) [inline efopen/newrelic_executor.py:120]: map_function should not be in this for loop as it is not related to page.

File: efopen/newrelic_executor.py [hunk 2]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
+      distribution_list = page['DistributionList']
```
Comment (csalagean-ellation) [inline efopen/newrelic_executor.py:121]: `distribution_list` variable is redundant.

File: efopen/newrelic_executor.py [hunk 3]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
+      filtered_distribution_list = []
+      for distribution in distribution_list['Items']:
+        response = cloudfront.list_tags_for_resource(distribution['ARN'])
+        # Remove from distribution list those that do not have tag nr_monitoring set to enabled
+        for tag in response['Tags']['Items']:
+          if tag['Key'].lower() == "nr_monitoring" and tag['Value'].lower() == "enabled":
+            filtered_distribution_list.append(distribution)
+            break
+      distribution_list['Items'] = filtered_distribution_list
       queue.extend(map(map_function, distribution_list['Items']))
```
Comment (csalagean-ellation) [inline efopen/newrelic_executor.py:131]: As stated above, you can use `filtered_distrubtion_list` directly.

File: efopen/newrelic_executor.py [hunk 4]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
-
+    # It makes no sense to continue if there is no distribution to monitor
+    if not queue:
+      return
```
Comment (csalagean-ellation) [inline efopen/newrelic_executor.py:134]: I'll also add a log message here.

**Review by csalagean-ellation — CHANGES_REQUESTED** [review]

**Review by momoneko — CHANGES_REQUESTED** [review]

File: efopen/newrelic_executor.py (hunk 4 above)
Comment (cristi-crunchyroll) [reply efopen/newrelic_executor.py:133]: Done

File: efopen/newrelic_executor.py [hunk 5]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
     logger.info("update cloudfront alert policy")
     cloudfront = self.clients['cloudfront']
-    distribution_list = cloudfront.list_distributions()['DistributionList']
-    map_function = lambda x: (x['Id'], ', '.join(x['Aliases']['Items']))
-    queue = map(map_function, distribution_list['Items'])
-
-    while distribution_list['IsTruncated']:
-      distribution_list = cloudfront.list_distributions(Marker=distribution_list['NextMarker'])['DistributionList']
+    paginator = cloudfront.get_paginator('list_distributions')
+    pages = paginator.paginate()
```
[the remark this answers is not in the document]
Comment (cristi-crunchyroll) [reply efopen/newrelic_executor.py]: Agreed

File: efopen/newrelic_executor.py [hunk 6]
```diff
@@ -113,14 +113,25 @@ def update_cloudfront_policy(self):
...
+    queue = []
+    for page in pages:
+      map_function = lambda x: (x['Id'], ', '.join(x['Aliases']['Items']))
+      distribution_list = page['DistributionList']
+      filtered_distribution_list = []
+      for distribution in distribution_list['Items']:
+        response = cloudfront.list_tags_for_resource(distribution['ARN'])
+        # Remove from distribution list those that do not have tag nr_monitoring set to enabled
+        for tag in response['Tags']['Items']:
+          if tag['Key'].lower() == "nr_monitoring" and tag['Value'].lower() == "enabled":
```
[the remark this answers is not in the document]
Comment (cristi-crunchyroll) [reply efopen/newrelic_executor.py]: Done

File: efopen/newrelic_executor.py [hunk 7]
```diff
@@ -113,13 +113,20 @@ def update_cloudfront_policy(self):
...
-    distribution_list = cloudfront.list_distributions()['DistributionList']
+    pages = cloudfront.get_paginator('list_distributions').paginate()
+    queue = []
     map_function = lambda x: (x['Id'], ', '.join(x['Aliases']['Items']))
-    queue = map(map_function, distribution_list['Items'])
-
-    while distribution_list['IsTruncated']:
-      distribution_list = cloudfront.list_distributions(Marker=distribution_list['NextMarker'])['DistributionList']
-      queue.extend(map(map_function, distribution_list['Items']))
+    pages = cloudfront.get_paginator('list_distributions').paginate()
```
Comment (momoneko) [inline efopen/newrelic_executor.py:119]: I think you have a duplicate line here. Duplicates 116
```suggestion
```

File: efopen/newrelic_executor.py [hunk 8]
```diff
@@ -113,13 +113,20 @@ def update_cloudfront_policy(self):
...
+    filtered_4xx_distributions = []
+    for page in pages:
+      for distribution in page['DistributionList']['Items']:
+        tag_response = cloudfront.list_tags_for_resource(Resource=distribution['ARN'])
+        # Remove from distribution list those that do not have tag nr_monitoring set to enabled
+        for tag in tag_response['Tags']['Items']:
+          if tag['Key'].lower() == "newrelic_4xx" and tag['Value'].lower() == "enabled":
```
Comment (momoneko) [inline efopen/newrelic_executor.py:126]: I probably worded this the wrong way during the last review.

Let's keep monitoring in the tag name.
```suggestion
          if tag['Key'].lower() == "newrelic_4xx_monitoring" and tag['Value'].lower() == "enabled":
```

**Review by momoneko — APPROVED** [review]

**Review by csalagean-ellation — APPROVED** [review]

**Pull request merged**